sql/select.go

```diff
@@ -85,3 +99,509 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
+// less than end. Note that whether the endpoints are inclusive or exclusive is
+// determined by qvalueInfo.op.
+type qvalueRange struct {
+	start qvalueInfo
+	end   qvalueInfo
+}
+
+type qvalueRangeMap map[ColumnID]*qvalueRange
+
+// analyzeExpr walks over the specified expression and populates the range map
+// with the value ranges for each qvalue.
+func (m qvalueRangeMap) analyzeExpr(expr parser.Expr) {
+	switch t := expr.(type) {
+	case *parser.NotExpr:
+		// TODO(pmattis): Similar to OR expressions, we can compute the value range
+		// for the expression and then invert the results.
+
+	case *parser.OrExpr:
+		// Conjunctions are handled below (see *parser.AndExpr). Disjunctions are
+		// handled by creating separate range maps for the left and right sides and
+		// then combining them. For example:
+		//
+		//   "x < y OR x < z"   -> "x < MAX(y, z)"
+		//   "x = y OR x = z"   -> "x >= y OR x <= z"
+		//   "x < y OR x > z"   -> "t"
+		left := make(qvalueRangeMap)
+		left.analyzeExpr(t.Left)
+		right := make(qvalueRangeMap)
+		right.analyzeExpr(t.Right)
+
+		for id, lRange := range left {
+			rRange, ok := right[id]
+			if !ok {
+				continue
+			}
+			r := &qvalueRange{}
+			if lRange.start.datum != nil && rRange.start.datum != nil {
+				r.start = lRange.start
+				r.start.union(rRange.start, true)
+			}
+			if lRange.end.datum != nil && rRange.end.datum != nil {
+				r.end = lRange.end
+				r.end.union(rRange.end, false)
+			}
+			if r.start.datum != nil || r.end.datum != nil {
+				m[id] = r
+			}
+		}
+
+	case *parser.AndExpr:
+		m.analyzeExpr(t.Left)
+		m.analyzeExpr(t.Right)
+
+	case *parser.ParenExpr:
+		m.analyzeExpr(t.Expr)
+
+	case *parser.ComparisonExpr:
+		m.analyzeComparisonExpr(t)
+
+	case *parser.RangeCond:
+		if t.Not {
+			// "a NOT BETWEEN b AND c" -> "a < b OR a > c"
+			m.analyzeExpr(&parser.OrExpr{
+				Left: &parser.ComparisonExpr{
+					Operator: parser.LT,
+					Left:     t.Left,
+					Right:    t.From,
+				},
+				Right: &parser.ComparisonExpr{
+					Operator: parser.GT,
+					Left:     t.Left,
+					Right:    t.To,
+				},
+			})
+		} else {
+			// "a BETWEEN b AND c" -> "a >= b AND a <= c"
+			m.analyzeExpr(&parser.AndExpr{
+				Left: &parser.ComparisonExpr{
+					Operator: parser.GE,
+					Left:     t.Left,
+					Right:    t.From,
+				},
+				Right: &parser.ComparisonExpr{
+					Operator: parser.LE,
+					Left:     t.Left,
+					Right:    t.To,
+				},
+			})
+		}
+	}
+}
+
+// analyzeComparison analyzes the comparison expression, restricting the start
+// and end info in indexInfo.elems if possible.
+func (m qvalueRangeMap) analyzeComparisonExpr(node *parser.ComparisonExpr) {
+	op := node.Operator
+	switch op {
+	case parser.EQ, parser.LT, parser.LE, parser.GT, parser.GE:
+		break
+
+	default:
+		// TODO(pmattis): For parser.LIKE we could extract the constant prefix and
+		// treat as a range restriction.
+		return
+	}
+
+	qval, ok := node.Left.(*qvalue)
```

Comment:
Right. There is nothing that ensures that the value is on the left. Might be worth doing in order to make a minor simplification here.